File: apps/server/app/services/enrollment.py
from typing import List
from fastapi import HTTPException, status

from crud.enrollment import EnrollmentCrud
from schemas.enrollment import (
    Enrollment as EnrollmentSchema,
    EnrollmentCreate,
    EnrollmentUpdate,
    EnrollmentResponse,
)
from schemas.student import Student as StudentSchema
from services.classroom import ClassroomService
from services.student import StudentService


class EnrollmentService:
    def __init__(
        self,
        crud: EnrollmentCrud,
        student_service: StudentService,
        classroom_service: ClassroomService,
    ) -> None:
        self.crud = crud
        self.student_service = student_service
        self.classroom_service = classroom_service

    async def create_enrollment(self, schema: EnrollmentCreate, instructor_id: int) -> EnrollmentResponse:
        # 0. 대상 반이 본인 소유인지 확인 (없거나 소유가 아니면 404)
        await self.classroom_service.get_swim_class_detail(schema.class_id, instructor_id)

        # 1. 학생 매칭/생성 (StudentService의 책임)
        student = await self.student_service.find_or_create_student(
            StudentSchema(
                name=schema.name,
                gender=schema.gender,
                phone=schema.phone,
                birth_year=schema.birth_year,
            )
        )

        # 2. Enrollment 생성 데이터 구성
        enrollment_data = {
            "student_id": student.id,
            "class_id": schema.class_id,
            "memo": schema.memo,
        }

        # 3. swim_class의 student_count 증가
        await self.classroom_service.increment_student_count(schema.class_id)

        # 4. Enrollment 생성
        enrollment = await self.crud.create(enrollment_data=enrollment_data)

        # 5. student + enrollment 반환
        return EnrollmentResponse(
            student=StudentSchema.model_validate(student),
            enrollment=EnrollmentSchema.model_validate(enrollment),
        )

    async def update_enrollment(
        self, enrollment_id: int, schema: EnrollmentUpdate, instructor_id: int
    ) -> EnrollmentResponse:
        # 1. 수정할 수강 정보 조회
        enrollment = await self.crud.get_by_id(enrollment_id)
        if not enrollment:
            raise HTTPException(
                status_code=status.HTTP_404_NOT_FOUND,
                detail="수강 신청 정보를 찾을 수 없습니다.",
            )

        # 1-1. 소속 반이 본인 소유인지 확인
        await self.classroom_service.get_swim_class_detail(enrollment.class_id, instructor_id)

        # 2. 수강 정보 업데이트
        updated_enrollment = await self.crud.update(enrollment, schema)

        # 3. student 정보와 함께 반환
        return EnrollmentResponse(
            student=StudentSchema.model_validate(updated_enrollment.student),
            enrollment=EnrollmentSchema.model_validate(updated_enrollment),
        )

    async def delete_enrollment(self, enrollment_id: int, instructor_id: int) -> bool:
        # 1. 수강 정보 존재 여부 확인
        enrollment = await self.crud.get_by_id(enrollment_id)
        if not enrollment:
            raise HTTPException(
                status_code=status.HTTP_404_NOT_FOUND,
                detail="수강 신청 정보를 찾을 수 없습니다.",
            )

        # 1-1. 소속 반이 본인 소유인지 확인
        await self.classroom_service.get_swim_class_detail(enrollment.class_id, instructor_id)

        # 2. swim_class의 student_count 감소
        await self.classroom_service.decrement_student_count(enrollment.class_id)

        # 3. Soft Delete 수행
        await self.crud.delete(enrollment_id)
        return True

    async def get_enrollments_by_class(
        self, class_id: int, instructor_id: int
    ) -> List[EnrollmentResponse]:
        # 0. 대상 반이 본인 소유인지 확인
        await self.classroom_service.get_swim_class_detail(class_id, instructor_id)

        # 특정 클래스의 전체 수강 목록 조회
        enrollments = await self.crud.get_by_class_id(class_id)
        return [
            EnrollmentResponse(
                student=StudentSchema.model_validate(e.student),
                enrollment=EnrollmentSchema.model_validate(e),
            )
            for e in enrollments
        ]

File: apps/mobile/src/api/enrollment.ts
import { GenderType } from '../types/member';
import { apiClient } from './client'; // 실제 경로에 맞게 조정 필요

// student_id는 백엔드가 name/phone/birth_year 기준으로 내부에서 매칭/생성한다고 가정
// -> 백엔드 EnrollmentCreate 스키마에서 student_id를 제거하거나 optional로 바꿔야 함 (확인 필요)
export interface EnrollmentCreate {
  class_id: number;
  name: string;
  gender?: GenderType;
  phone?: string;
  birth_year?: number;
  memo?: string;
}

export interface EnrollmentStudent {
  id: number;
  name: string;
  gender?: GenderType;
  phone?: string;
  birth_year?: number;
}

export interface EnrollmentDetail {
  id: number;
  student_id: number;
  class_id: number;
  memo?: string;
  created_at: string;
  deleted_at: string | null;
}

export interface EnrollmentResponse {
  student: EnrollmentStudent;
  enrollment: EnrollmentDetail;
}

export const enrollmentApi = {
  createEnrollment: async (data: EnrollmentCreate) => {
    const response = await apiClient.post<EnrollmentResponse>('/enrollment', data);
    console.log(response)
    return response.data;
  },

  deleteEnrollment: async (enrollmentId: number) => {
    const response = await apiClient.delete(`/enrollment/${enrollmentId}`);
    console.log(response);
    return response.data;
  },

  updateEnrollment: async (enrollmentId: number, data: EnrollmentDetail) => {
    const response = await apiClient.patch(`/enrollment/${enrollmentId}`, data);
    console.log(response);
    return response.data;
  },

  getEnrollments: async (classId: number) => {
    const response = await apiClient.get<EnrollmentResponse[]>(`/enrollment/class/${classId}`);
    console.log(response)
    return response.data;
  },
};
